From b31de7d210fcde578a61f08c8df72a510bc0a843 Mon Sep 17 00:00:00 2001
From: "Yong, Jonathan" <jonathan.yong@intel.com>
Date: Wed, 24 Aug 2016 04:36:17 +0000
Subject: [PATCH 11/41] Prevent sbi_apl from crashing due to race condition

commit 402bda389a665597fd4ea49c20df5045c551d747 from
git://git.yoctoproject.org/linux-yocto-4.1

Avoids null dereference.
The driver may be called before it is fully initialized.

Signed-off-by: Yong, Jonathan <jonathan.yong@intel.com>
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 arch/x86/platform/bxt/sbi_apl.c | 19 ++++++++++---------
 1 file changed, 10 insertions(+), 9 deletions(-)

diff --git a/arch/x86/platform/bxt/sbi_apl.c b/arch/x86/platform/bxt/sbi_apl.c
index 1974f5faaba0..46e0a6f1395b 100644
--- a/arch/x86/platform/bxt/sbi_apl.c
+++ b/arch/x86/platform/bxt/sbi_apl.c
@@ -31,9 +31,6 @@
 #define DRV_VERSION "1.0"
 
 static struct sbi_platform_data *plat_data;
-#define sbi_pdev_bus plat_data->bus
-#define sbi_pdev_slot plat_data->p2sb
-#define sbi_apl_lock plat_data->lock
 
 static u32 sbi_address(const struct sbi_apl_message *args)
 {
@@ -169,18 +166,22 @@ int sbi_apl_commit(struct sbi_apl_message *args)
 	int ret;
 	struct pci_bus *sbi_pdev;
 
+	/* We're called before we're ready */
+	if (!plat_data)
+		return -EAGAIN;
+
 	ret = sbi_validate(args);
 	if (ret)
 		return ret;
 
-	sbi_pdev = pci_find_bus(0, sbi_pdev_bus);
+	sbi_pdev = pci_find_bus(0, plat_data->bus);
 	if (!sbi_pdev)
 		return -ENODEV;
-	mutex_lock(sbi_apl_lock);
-	sbi_hide(sbi_pdev, sbi_pdev_slot, 0);
-	ret = sbi_do_write(sbi_pdev, sbi_pdev_slot, args);
-	sbi_hide(sbi_pdev, sbi_pdev_slot, 1);
-	mutex_unlock(sbi_apl_lock);
+	mutex_lock(plat_data->lock);
+	sbi_hide(sbi_pdev, plat_data->p2sb, 0);
+	ret = sbi_do_write(sbi_pdev, plat_data->p2sb, args);
+	sbi_hide(sbi_pdev, plat_data->p2sb, 1);
+	mutex_unlock(plat_data->lock);
 
 	return ret;
 }
-- 
2.7.4

